apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,122 @@
+// idempotency.interceptor.ts
+
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  Logger,
+  UnprocessableEntityException,
+  HttpException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 24; //24h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5m
+
+enum ReqStatusEnum {
+  PROGRESS = 'in-progress',
+  SUCCESS = 'success',
+  ERROR = 'error',
+}
+
+@Injectable()
+export class IdempotencyInterceptor implements NestInterceptor {
+  constructor(private readonly cacheService: CacheService) {}
+
+  async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
+    const request = context.switchToHttp().getRequest();
+    const idempotencyKey = request.headers['idempotency-key'];
+    if (!idempotencyKey || !['post', 'patch'].includes(request.method.toLowerCase())) {
+      return next.handle();
+    }
+    const { organizationId } = request.user || {};
+    const env = process.env.NODE_ENV;
+    const cacheKey = `${env}-${organizationId}-${idempotencyKey}`;
+
+    try {
+      const data = await this.cacheService.get(cacheKey);
+      // Check if the idempotency key is in the cache
+      if (!data || !JSON.parse(data)?.status) {
+        //1st time we are seeing the request mark the request as in-progress
+        await this.setCache(cacheKey, { status: ReqStatusEnum.PROGRESS }, IDEMPOTENCY_PROGRESS_TTL);
+
+        return next.handle().pipe(
+          map(async (response) => {
+            const statusCode = context.switchToHttp().getResponse().statusCode;
+            // Cache the success response and return it
+            await this.setCache(
```

Comment:
I think we should add a verbose logger call here that says we are caching this response.